File: TotalCrossVM/src/util/tcz.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

#include "tcvm.h"
#include "tcz.h"

#ifdef ANDROID // in Android, we use Java methods to read directly from the apk
int32 callFindTCZ(CharP name)
{
   JNIEnv* env = getJNIEnv();
   jstring jname = (*env)->NewStringUTF(env, (const char*)name);
   int32 ret = (*env)->CallStaticIntMethod(env, applicationClass, jfindTCZ, jname);
   (*env)->DeleteLocalRef(env, jname);
   return ret;
}

// public static int loadDataFromAPK(String file, int ofs, byte[] buf)
static int32 callReadTCZ(int32 apkIdx, uint8* buf, int32 offset, int32 length)
{
   JNIEnv* env = getJNIEnv();
   jbyteArray jbytesP = (*env)->NewByteArray(env, length); // !!! temporary byte array has length: count-offset
   int32 ret,i;

   ret = (*env)->CallStaticIntMethod(env, applicationClass, jreadTCZ, apkIdx, offset, jbytesP);
   
   if (ret > 0)
   {                 
      jbyte* jbytes = (*env)->GetByteArrayElements(env, jbytesP, 0); // android5 require get the array after returning from method
      xmemmove(buf, jbytes, ret);
      (*env)->ReleaseByteArrayElements(env, jbytesP, jbytes, 0);
   }
   
   (*env)->DeleteLocalRef(env, jbytesP);
   return ret;
}
#endif

/**
    A tcz file has the following format:
    . version
    . attributes
    . length
    . size of compressed offsets
    . size of compressed names
    . compressed offsets array (length+1)
    . uncompressed sizes array (length)
    . compressed names array (length)
    . compressed data chunks

    The first record is the class that implements totalcross.MainClass.
*/

void destroyTCZ() // no threads are running at this point
{
   while (openTCZs != null)  // the tcz is removed from the list on tczClose
      tczClose((TCZFile)openTCZs->value);
}

static bool tczReadMore(TCZFile f)
{
   int32 n;
   bool ret = true;
   LOCKVAR(tcz);
#ifdef ANDROID
   n = callReadTCZ(f->header->apkIdx, f->buf, f->expectedFilePos, TCZ_BUFFER_SIZE);
#else   
   if (f->expectedFilePos != f->header->realFilePos) // if the file position has changed by some other instance, reposition it
      fseek(f->header->fin, f->expectedFilePos, SEEK_SET);
   n = (int32)fread(f->buf, 1, TCZ_BUFFER_SIZE, f->header->fin);
#endif   
   f->header->realFilePos = (f->expectedFilePos += n);
   if (n <= 0)
      ret = false; // no more data
   else
   {
      f->zs.next_in = f->buf;
      f->zs.avail_in = n;
   }
   UNLOCKVAR(tcz);
   return ret;
}

int32 tczRead(TCZFile f, void* outBuf, int32 count)
{
   int32 err=0;
   z_stream *zs = &f->zs;
   zs->avail_out = count;
   zs->next_out = outBuf;
   if (count == 0)
      return 0;

   while (true)
   {
      if (zs->avail_in == 0 && !tczReadMore(f))
         break;
      err = inflate(&f->zs, Z_NO_FLUSH);
      if (err == Z_STREAM_END)
      {
         int32 ret = count - zs->avail_out;
         inflateEnd(&f->zs);
         return ret;
      }
      else
      if (err < 0 || zs->avail_out == 0)
         break;
   }
   if (zs->avail_out > 0)
   {
      if (f->tempHeap != null)
         HEAP_ERROR(f->tempHeap, HEAP_ZIP_ERROR);
      else
         debug("Error on zip (in a heapless tcz): %d. Remain %d bytes",(int)err, (int)zs->avail_out);
   }
   return count;
}

int32 tczRead32(TCZFile f)
{
   int32 i=0;
   tczRead(f, &i, 4);
   return i;
}

int32 tczRead32BE(TCZFile f)
{
   int32 i=0;
   tczRead(f, &i, 4);
   i = SWAP32_FORCED(i);
   return i;
}

int16 tczRead16(TCZFile f)
{
   int16 i=0;
   tczRead(f, &i, 2);
   return i;
}

int16 tczRead16BE(TCZFile f)
{
   int16 i=0;
   tczRead(f, &i, 2);
   i = SWAP16_FORCED(i);
   return i;
}

int8 tczRead8(TCZFile f)
{
   int8 i=0;
   tczRead(f, &i, 1);
   return i;
}

#ifndef ANDROID
static void tczFinalizer(Heap heap, void* bag)
{
   TCZFileHeader header = (TCZFileHeader)bag;
   UNUSED(heap);
   fclose(header->fin);
   header->fin = null;
}
#endif

static TCZFile tczNewInstance(TCZFile parent)
{
   volatile Heap hheap = null;
   TCZFile ntcz;
   int32 err;
   ntcz = newX(TCZFile); // don't use the heap
   if (!ntcz)
      return null;
   if (parent) // not first instance?
   {
      ntcz->header = parent->header;
      IF_HEAP_ERROR(ntcz->header->hheap)
         goto error;
   }
   else
   {
      hheap = heapCreate();
      IF_HEAP_ERROR(hheap)
         goto error;
      ntcz->header = newXH(TCZFileHeader, hheap);
      ntcz->header->hheap = hheap;
#ifndef ANDROID
      heapSetFinalizer(hheap, tczFinalizer, ntcz->header);
#endif      
   }
   //debug("tczNewInstance tcz %d from header %d - %d",(int32)ntcz, (int32)ntcz->header, ntcz->header->instanceCount);
   ntcz->zs.opaque = ntcz->header->hheap;
   err = inflateInit(&ntcz->zs);
   if (err != Z_OK)
      goto error;
   ntcz->header->instanceCount++;
   return ntcz;
error:
   if (hheap) heapDestroy(hheap);
   xfree(ntcz);
   return null;
}

void tczClose(TCZFile tcz)
{
   if (tcz)
   {
      //debug("closing tcz %d from header %d - %d",(int32)tcz, (int32)tcz->header, tcz->header->instanceCount);
      inflateEnd(&tcz->zs);
      // remove the tcz from the list. Note that the first tcz added is usually the last one deleted; the exception to this is when we get an error while loading the constant pool.
      openTCZs = VoidPsRemove(openTCZs, tcz, null);
      if (--tcz->header->instanceCount == 0) // if there are no more instances, destroy the heap
      {
         // destroy the heaps
         if (tcz->header->cp)
            heapDestroy(tcz->header->cp->heap);
         heapDestroy(tcz->header->hheap);
      }
      xfree(tcz);
   }
}

static int32 findNamePosition(TCZFile tcz, CharP name)
{
   CharPArray names = tcz->header->names;
   if (strEq(names[0], name)) // is it the first file (extends MainClass)?
      return 0;
   else
   {
      int32 inf = 1,sup = ARRAYLEN(names)-1,half,res;
      while (inf<=sup)
      {
         half = (inf+sup)>>1;
         res = xstrcmp(name, names[half]);
         if (res == 0)
            return half;
         if (res < 0)
            sup=half-1;
         else
            inf=half+1;
      }
      return -1;
   }
}

TCZFile tczFindName(TCZFile tcz, CharP name) // locates the name and also positions the stream at the place to start reading it
{
   TCZFile ntcz = null;
   int32 pos;
   LOCKVAR(tcz);
   pos = findNamePosition(tcz, name);
   if (pos == -1)
      goto end;
   ntcz = tczNewInstance(tcz);
   if (!ntcz)
      goto end;
   ntcz->header->realFilePos = ntcz->expectedFilePos = ntcz->header->offsets[pos];
#ifndef ANDROID   
   fseek(ntcz->header->fin, ntcz->expectedFilePos, SEEK_SET);
#endif   
   ntcz->uncompressedSize = tcz->header->uncompressedSizes[pos];
end:
   UNLOCKVAR(tcz);
   return ntcz;
}

// Reads a TCZ file and place the informations in the public members available in this "class".
#ifdef ANDROID
TCZFile tczOpen(CharP fileName, bool isFont)
#else
TCZFile tczOpen(FILE* fin, CharP fileName)
#endif
{
   int32 baseOffset,n,i;
   CharPArray names;
   int32 *offsets;
   TCZFile tcz;
   int16 version, attr;
   volatile Heap heap;
   // assuming that current fin position is 0
#ifdef ANDROID
   uint8 header[8],*h = header;
   int32 apkIdx = callFindTCZ(fileName);
   if (apkIdx < 0)
      return null;
   n = callReadTCZ(apkIdx, h, 0, 8);
   if (n < 0)
      return null;
   xmemmove(&version,h,2); h += 2;
   xmemmove(&attr,h,2); h += 2;
   xmemmove(&baseOffset,h,4);
#else   
   version    = fread16(fin);
   attr       = fread16(fin);
   baseOffset = fread32(fin);
#endif   
   if (version == 0)
   {
      alert("Invalid TCZ version.\nFile probably corrupted");
      return null;
   }
   if (
#ifdef ANDROID
      !isFont
#else            
      fileName != null 
#endif      
      && version != TCZ_VERSION) // guich@tc110_70: check for version mismatch. font files doesn't care if the tcz changed
   {
      alert("TCZ version mismatch for %s (defined: %d, loaded: %d). Recompile and deploy your app with the new SDK.",fileName,TCZ_VERSION,version);
      return null;
   }
   // the instance can only be created after the initial part has been read
   tcz = tczNewInstance(null);
   if (!tcz)
      return null;

   tcz->header->realFilePos = tcz->expectedFilePos = 8;
#ifdef ANDROID
   tcz->header->apkIdx = apkIdx;
#else      
   tcz->header->fin = fin;
#endif   
   tcz->header->version = version;
   tcz->header->attr = attr;
   heap = tcz->tempHeap = tcz->header->hheap;
   IF_HEAP_ERROR(heap)
   {
      heapDestroy(heap);
      //alert("opentcz ERROR: %d\nat %s (%d)",heap->errorCode, heap->errorFile, heap->errorLine);
      tczClose(tcz);
      return null;
   }
   n = tczRead32(tcz);

   // allocate space for names and offsets
   names   = tcz->header->names   = newPtrArrayOf(CharP, n, heap);
   offsets = tcz->header->offsets = newPtrArrayOf(Int32,  n+1, heap);
   tcz->header->uncompressedSizes = newPtrArrayOf(Int32,  n, heap);
   // decompress the offsets
   tczRead(tcz, tcz->header->offsets, (n+1) * 4);
   tczRead(tcz, tcz->header->uncompressedSizes, n * 4);
   for (i = n; i >= 0; i--) // add the baseOffset
      *offsets++ += baseOffset;

   // decompress the names
   for (i = 0; i < n; i++, names++)
   {
      int32 len = (uint8)tczRead8(tcz);
      *names = (CharP)heapAlloc(heap, len+1);
      tczRead(tcz, *names, len);
   }
   tcz->tempHeap = null; // not necessary, but safe
   inflateEnd(&tcz->zs);
   return tcz; // do NOT close this tcz here! it will be closed later
}

extern void readConstantPool(Context currentContext, ConstantPool t, TCZFile tcz, Heap heap); // class.c

TCZFile tczLoad(Context currentContext, CharP tczName)
{
#ifndef ANDROID   
   FILE* f;    
#endif   
   volatile TCZFile t=null,t2=null;

#ifdef ANDROID
   if ((t = tczOpen(tczName, false)) != null)
#else      
   f = findFile(tczName,null);
   if (f != null && (t = tczOpen(f, tczName)) != null)
#endif      
   {
      VoidPs* temp;
      // enqueue the file in the TCZ list
      //debug("enqueuing tcz %d from header %d - %d",(int32)t, (int32)t->header, t->header->instanceCount);
      temp = VoidPsAdd(openTCZs, t, null); // cannot use heap here!
      if (temp == null)
      {
         tczClose(t);
         return null;
      }
      openTCZs = temp;
      // load the constant pool from inside this file
      if ((t2=tczFindName(t, "ConstantPool")) != null) // classes, not just resources?
      {
         volatile Heap cpHeap = heapCreate();
         IF_HEAP_ERROR(cpHeap)
         {                                                                 
            alert("Please run the program again (104)");
            debug("Error when loading constant pool of file\n%s", tczName);
            heapDestroy(cpHeap);
            t->header->cp = null; // already destroyed
            tczClose(t);
            tczClose(t2);
            return null;
         }
         t->header->cp = newXH(ConstantPool, cpHeap);
         readConstantPool(currentContext, t->header->cp, t2, cpHeap);
         tczClose(t2);
      }
   }
   return t;
}

TCZFile tczGetFile(CharP filename, bool strict) // fdie@ strict mode prevents mainClass package name based resolving
{
   TCZFile found = null;
   VoidPs *list, *head;
   list = head = openTCZs;
   if (head != null)
   do
   {
      TCZFile tcz = (TCZFile)list->value;
      found = tczFindName(tcz, filename);
      list = list->next;
   } while (head != list && found == null);

   // if the resource has not been found and "strict" mode is disabled, try to prefix the resource with the mainClass package name
   if (found == null && !strict)
   {
      char fullpath[MAX_PATHNAME];
      char *lastdot;
      xstrcpy(fullpath, mainClassName);
      lastdot = xstrrchr(fullpath, '.');
      if (lastdot != null)
      {
         lastdot[1] = 0;
         replaceChar(fullpath, '.', '/'); // now contains the package name with '/' separator
         xstrcpy(lastdot + 1, filename); // append the resource name and try again in "strict" mode
         found = tczGetFile(fullpath, true);
      }
   }

   return found;
}

File: TotalCrossVM/builders/droid/app/src/main/java/totalcross/android/fontgen/TCZ.java
// Copyright (C) 2000-2011 SuperWaba Ltda. 
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only


package totalcross.android.fontgen;

import totalcross.*;

import android.os.*;
import java.io.*;
import java.util.*;
import java.util.zip.*;

/**
    A tcz (TotalCross Zip) file has the following format:
    <ul>
    <li> version     (2 bytes)
    <li> attributes  (2 bytes)
    <li> base offset (header size + 4)
    <li> header
    <li> compressed data chunks
    </ul>

    The header is:

    <ul>
    <li> length      (4 bytes)
    <li> offsets array (length+1) - offset[i+1]-offset[i] gives the compressed size
    <li> uncompressed sizes array (length)
    <li> names array (length)
    </ul>

    The header is compressed to save space.
    The first record is the class that implements totalcross.MainClass or extends totalcross.ui.MainWindow.
*/

public class TCZ
{
   /** Set this to be the main class name. It is stored in record #0. */
   public static String mainClassName; // remove dependency on DeploySettings

   /** An entry of the TCZ file. */
   public static class Entry
   {
      /** The compressed byte block. */
      public byte[] bytes;
      /** The name of the entry. */
      public String name;
      /** The size of the block when it is uncompressed. */
      public int uncompressedSize;
      /** Anything you want to hold here. */
      public Object extra; // JavaClass for the converter
      
      private String name2write; // guich@tc115_23

      public Entry(byte []bytes, String name, int uncompressedSize) throws Exception
      {
         this(bytes, name, uncompressedSize, null);
      }

      public Entry(byte []bytes, String name, int uncompressedSize, Object extra) throws Exception
      {
         this.uncompressedSize = uncompressedSize;
         this.bytes = bytes;
         this.name = name.replace('\\','/');
         this.extra = extra;
         if (name.endsWith(".class"))
            name = name.substring(0,name.length()-6);
         name2write = name;
         if (name2write.indexOf('.') < 0) // no dots on string? note that an image will have a dot, so its correctly kept with / on the name
            name2write = name2write.replace('/','.'); // assuming not ending with .class
      }

      /** Returns a String representing this Entry. Used in the Vector.qsort method */
      public String toString()
      {
         if (mainClassName != null && name.equals(mainClassName))
            return '\1'+name2write; // make sure this name will go first when sorting
         return name2write;
      }
   }

   public static final short TCZ_VERSION = 106; // must sync with tcz.h

   /** Defines that the tcz file has a MainClass at the first record. If false, it is a library-only module */
   public static final short ATTR_HAS_MAINCLASS = 1;  
   /** Defines that the tcz file has a MainWindow at the first record. */
   public static final short ATTR_HAS_MAINWINDOW = 2;
   /** Defines that the tcz file is a library-only module. */  
   public static final short ATTR_LIBRARY = 4;

   /** The names of the files. */
   public String[] names;
   /** The attributes.
    * @see #ATTR_HAS_MAINCLASS
    * @see #ATTR_HAS_MAINWINDOW
    * @see #ATTR_LIBRARY
    */
   public int attr;
   /** Version of the tcz file.
    * @see #TCZ_VERSION
    */
   public int version;
   /** Offsets to the compressed data chunks */
   public int[] offsets;
   /** Sizes of the data chunks when they are uncompressed. */
   public int[] uncompressedSizes;
   /** The number of chunks. */
   public int numberOfChunks;
   /** Stores the total size when a TCZ file is created. */
   public int size;
   /** Bag that can be used to store anything that the user wants. */
   public Object bag;

   /** Create a TCZ file with the given vector of Entry(s).
    * @throws IOException
    * @throws ZipException */
   public TCZ(Vector<Entry> vout, String outName, short attr) throws IOException
   {
      if (!outName.toLowerCase().endsWith(".tcz"))
         outName =  outName.concat(".tcz");
      // creates an empty file for output
      String ss = Environment.getExternalStorageDirectory()+"/"+"_"+outName;
      FileOutputStream fout = new FileOutputStream(ss);
      
      Object[] out = vout.toArray();
      sort(out,0,out.length-1);

      // now we process the files.
      int n = out.length;

      // first pass, we setup the names and offset arrays
      offsets = new int[n+1]; // first offset is 0
      names = new String[n];
      uncompressedSizes = new int[n];
      int ofs = 0;
      for (int i =0; i < n; i++)
      {
         Entry of = (Entry)out[i];
         names[i] = of.name2write;
         ofs += of.bytes.length;
         offsets[i+1] = ofs;
         uncompressedSizes[i] = of.uncompressedSize;
      }

      // prepare the header
      ByteArrayOutputStream header = new ByteArrayOutputStream(4096);
      
      writeInt(header,n);
      for (int i = 0; i <= n; i++)
         writeInt(header,offsets[i]); // the offsets are stored first to ensure alignment.
      for (int i = 0; i < n; i++)
         writeInt(header,uncompressedSizes[i]); // idem
      for (int i = 0; i < n; i++) // strings are stored as UTF8
      {
         String s = names[i];
         int l = s.length();
         if (l > 255) throw new ZipException("Error: Name too long: "+s+". ("+l+" chars). Maximum allowed: 255 chars");
         writeSmallString(header,s);
      }
      byte[] hh = header.toByteArray();
      byte[] compressedHeaderBytes = compress(hh);
      AndroidUtils.debug("header: "+hh.length+" -> "+compressedHeaderBytes.length);

      // write the first part: the version, the attribute and the compressed header
      size += writeShort(fout, TCZ_VERSION);
      size += writeShort(fout, attr);
      size += writeInt(fout, compressedHeaderBytes.length+8); // base offset = compressed header size + 4
      fout.write(compressedHeaderBytes, 0, compressedHeaderBytes.length);
      size += compressedHeaderBytes.length;
      // now write the compressed chunks
      for (int i = 0; i < n; i++)
      {
         byte[] bytes = ((Entry)out[i]).bytes;
         fout.write(bytes);
         size += bytes.length;
      }
      fout.close();
      AndroidUtils.debug("OUTPUT: "+ss);
   }
   
   public static byte[] compress(byte[] in) throws IOException
   {
      ByteArrayOutputStream bc = new ByteArrayOutputStream(in.length/2);
      Deflater dd = new Deflater(9, false);
      DeflaterOutputStream def = new DeflaterOutputStream(bc, dd);
      def.write(in);
      def.finish();
      dd.end();
      return bc.toByteArray();
   }

   private static void sort(Object []items, int first, int last)
   {
      if (first >= last)
         return;
      int low = first;
      int high = last;

      String mid = items[(first+last) >> 1].toString();
      while (true)
      {
         while (high >= low && mid.compareTo(items[low].toString())  > 0) // guich@566_25: added "high > low" here and below - guich@568_5: changed to >=
            low++;
         while (high >= low && mid.compareTo(items[high].toString()) < 0)
            high--;
         if (low <= high)
         {
            Object temp = items[low];
            items[low++] = items[high];
            items[high--] = temp;
         }
         else break;
      }

      if (first < high)
         sort(items,first,high);
      if (low < last)
         sort(items,low,last);
   }

   public static int writeShort(OutputStream stream, int i) throws IOException
   {
      byte[] b = buffer;
      b[0] = (byte)i;
      i >>= 8;
      b[1] = (byte)i;
      stream.write(b, 0, 2);
      return 2;
   }

   static private byte[] buffer = new byte[256]; // starts with 256 bytes since readSmallString uses it
   
   public static int writeSmallString(OutputStream stream, String s) throws IOException
   {
      int len = s == null ? 0 : s.length();
      if (s.length() > 255) throw new IOException("String size "+s.length()+" is too big to use with writeSmallString!");
      int pos = 0;
      buffer[pos++] = (byte)len;
      int ret = len+1;
      if (len > 0)
         for (int i = 0; len-- > 0; i++)
            buffer[pos++] = (byte)s.charAt(i);
      stream.write(buffer, 0, pos);
      return ret;
   }

   public static int writeInt(OutputStream stream, int i) throws IOException
   {
      byte[] b = buffer;
      b[0] = (byte)i;
      i >>= 8; // guich@300_40
      b[1] = (byte)i;
      i >>= 8;
      b[2] = (byte)i;
      i >>= 8;
      b[3] = (byte)i;
      stream.write(b, 0, 4);
      return 4;
   }
}
